impl/src/main/java/org/jboss/weld/xml/BeansXmlStreamParser.java

```diff
@@ -0,0 +1,398 @@
+    public static final String JAVAEE_URI = "http://xmlns.jcp.org/xml/ns/javaee";
+    public static final Set<String> JAVAEE_URIS = ImmutableSet.of(JAVAEE_LEGACY_URI, JAVAEE_URI);
+
+    public static final String WELD_URI = "http://jboss.org/schema/weld/beans";
+    public static final Set<String> SCANNING_URIS = ImmutableSet.of(WELD_URI, JAVAEE_URI, JAVAEE_LEGACY_URI);
+
+    private static final String VERSION_ATTRIBUTE_NAME = "version";
+    private static final String BEAN_DISCOVERY_MODE_ATTRIBUTE_NAME = "bean-discovery-mode";
+    private static final String NAME_ATTRIBUTE_NAME = "name";
+    private static final String VALUE_ATTRIBUTE_NAME = "value";
+    private static final String PATTERN_ATTRIBUTE_NAME = "pattern";
+
+    private static final String IF_CLASS_AVAILABLE = "if-class-available";
+    private static final String IF_CLASS_NOT_AVAILABLE = "if-class-not-available";
+    private static final String IF_SYSTEM_PROPERTY = "if-system-property";
+
+    private static final String CLASS = "class";
+    private static final String STEREOTYPE = "stereotype";
+    private static final String INCLUDE = "include";
+    private static final String EXCLUDE = "exclude";
+    private static final String TRIM = "trim";
+    private static final String BEANS = "beans";
+    private static final String ALTERNATIVES = "alternatives";
+    private static final String INTERCEPTORS = "interceptors";
+    private static final String DECORATORS = "decorators";
+    private static final String SCAN = "scan";
+
+    private List<Metadata<String>> enabledInterceptors = null;
+    private List<Metadata<String>> enabledDecorators = null;
+    private List<Metadata<String>> selectedAlternatives = null;
+    private List<Metadata<String>> selectedAlternativeStereotypes = null;
+    private List<Metadata<Filter>> includes = null;
+    private List<Metadata<Filter>> excludes = null;
+    private BeanDiscoveryMode discoveryMode = BeanDiscoveryMode.ALL;
+    private String version;
+    private boolean isTrimmed;
+
+    private final URL beansXml;
+
+    private final Function<String, String> interpolator;
+
+    /**
+     *
+     * @param beansXml
+     */
+    public BeansXmlStreamParser(URL beansXml) {
+        this(beansXml, Function.identity());
+    }
+
+    /**
+     *
+     * @param beansXml
+     * @param interpolator
+     */
+    public BeansXmlStreamParser(URL beansXml, Function<String, String> interpolator) {
+        this.beansXml = beansXml;
+        this.interpolator = interpolator;
+    }
+
+    public BeansXml parse() {
+        if (beansXml == null) {
+            throw XmlLogger.LOG.loadError("unknown", null);
+        }
+        try (InputStream in = beansXml.openStream()) {
+            if (in.available() == 0) {
+                // The file is just acting as a marker file
+                return EMPTY_BEANS_XML;
+            }
+            XMLInputFactory factory = XMLInputFactory.newInstance();
```

Comment:
Maybe share this?  Repeated creation of the factory can be very expensive.

Reply:
Well, the API does not explicitly declare whether it's ok to share `XMLInputFactory`. But it seems that it should be safe. I can try to run some tests and we can see if it's worth the risk.

Reply:
No, it's not safe, but you're talking about a one-line `synchronized` section, e.g.:

    XMLEventReader reader;
    synchronized (XML_INPUT_FACTORY) {
        reader = XML_INPUT_FACTORY.createXMLEventReader(in);
    }

I'm pretty sure that in a multi-`beans.xml` environment you're going to see a pretty good performance improvement from this; I'll move on to profiling later today.  Thanks for your work on all this!

Reply:
Oh, plus this will let us use Woodstox, which will _really_ speed things up!

Reply:
I think that Woodstox is an impl of StAX. It's up to the environment /user which impl is used.